File: work-engine/src/docs/docRegistry.ts
/**
 * Document registry.
 *
 * Ports `lambda-functions/src/lambda_functions/doc_registry.py` to TypeScript:
 * builds a registry of `content/` markdown documents with stable IDs, aliases,
 * and path lookups, and resolves `doc:` / wiki / path references. Frontmatter is
 * parsed with the SOP frontmatter reader (the same YAML subset the Python
 * registry used via `sop_parse.parse_frontmatter`).
 */

import { readdirSync, readFileSync, statSync } from 'node:fs';
import { join, posix, sep } from 'node:path';

import { parseFrontmatter, splitFrontmatter } from './sop';
import type { Frontmatter } from './sop';

const DOCUMENT_ID_RE = /^[a-z0-9][a-z0-9._-]*$/;
const WIKI_REF_RE = /^\[\[([^\]|]+)(?:\|[^\]]+)?\]\]$/;
const ID_STRIP_DIRS = new Set(['sops', 'templates', 'reference', 'playbooks']);

export const VALID_DOC_TYPES = new Set([
  'sop',
  'checklist',
  'template',
  'reference',
  'playbook',
  'prompt',
  'task-template',
  'archive',
  'doc',
]);

/** Thrown when registry validation finds violations. */
export class DocumentRegistryError extends Error {
  readonly violations: string[];
  constructor(violations: string[]) {
    super(violations.join('\n'));
    this.name = 'DocumentRegistryError';
    this.violations = violations;
  }
}

export interface DocumentRecord {
  id: string;
  aliases: string[];
  path: string;
  title: string;
  doc_type: string;
  summary: string;
  tags: string[];
  systems: string[];
  related_docs: string[];
  updated_at: number;
  domain: string;
  id_source: 'frontmatter' | 'generated';
}

export interface DocumentRecordDict extends Omit<DocumentRecord, never> {
  stable_id: boolean;
}

export interface DocumentRegistry {
  documents: DocumentRecord[];
  byId: Map<string, DocumentRecord>;
  byAlias: Map<string, DocumentRecord>;
  byPath: Map<string, DocumentRecord>;
}

export function recordToDict(record: DocumentRecord): DocumentRecordDict {
  return {
    id: record.id,
    stable_id: record.id_source === 'frontmatter',
    aliases: record.aliases,
    path: record.path,
    title: record.title,
    doc_type: record.doc_type,
    summary: record.summary,
    tags: record.tags,
    systems: record.systems,
    related_docs: record.related_docs,
    updated_at: record.updated_at,
    domain: record.domain,
    id_source: record.id_source,
  };
}

export function registryToDict(registry: DocumentRegistry): { documents: DocumentRecordDict[] } {
  return { documents: registry.documents.map(recordToDict) };
}

// ── Build ─────────────────────────────────────────────────────────────────────

/** Recursively collect `*.md` files under `dir`, sorted (posix order). */
function collectMarkdown(dir: string): string[] {
  const out: string[] = [];
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) out.push(...collectMarkdown(full));
    else if (entry.isFile() && entry.name.endsWith('.md')) out.push(full);
  }
  return out;
}

/**
 * Build the registry for a content root. `contentRoot` is an absolute path; repo
 * paths are computed against its parent (the repo/cache root), matching Python
 * `build_registry`.
 */
export function buildRegistry(contentRoot: string, validate = true): DocumentRegistry {
  const repoRoot = posixParent(contentRoot);
  const records: DocumentRecord[] = [];
  const violations: string[] = [];

  const files = collectMarkdown(contentRoot).sort();
  for (const file of files) {
    const text = readFileSync(file, 'utf-8');
    const [rawFm, body] = splitFrontmatter(text);
    const metadata: Frontmatter = rawFm ? parseFrontmatter(rawFm) : {};
    const repoPath = toPosixRel(file, repoRoot);
    const mtime = Math.floor(statSync(file).mtimeMs / 1000);
    const record = recordFromMetadata(repoPath, body, metadata, mtime);
    records.push(record);

    if (!DOCUMENT_ID_RE.test(record.id)) violations.push(`${repoPath}: invalid id '${record.id}'`);
    for (const alias of record.aliases) {
      if (looksLikePath(alias)) {
        try {
          normalizePathAlias(alias);
        } catch {
          violations.push(`${repoPath}: invalid path alias '${alias}'`);
        }
      } else if (!DOCUMENT_ID_RE.test(alias)) {
        violations.push(`${repoPath}: invalid alias '${alias}'`);
      }
    }
    if (!VALID_DOC_TYPES.has(record.doc_type)) {
      violations.push(`${repoPath}: unsupported doc_type '${record.doc_type}'`);
    }
  }

  const byId = new Map<string, DocumentRecord>();
  const byAlias = new Map<string, DocumentRecord>();
  const byPath = new Map<string, DocumentRecord>();

  for (const record of records) {
    addUnique(byId, record.id, record, 'id', violations);
    byPath.set(record.path, record);
    byPath.set(removePrefix(record.path, 'content/'), record);
    byPath.set('/' + removePrefix(record.path, 'content/'), record);
    for (const alias of record.aliases) {
      addUnique(byAlias, normalizeAlias(alias), record, 'alias', violations);
    }
  }

  const registry: DocumentRegistry = { documents: records, byId, byAlias, byPath };

  if (validate) {
    violations.push(...validateAliasConflicts(byAlias, byId));
    violations.push(...validateAliasPathConflicts(byAlias, byPath));
    violations.push(...validateRelatedDocs(records, registry));
    if (violations.length) throw new DocumentRegistryError(violations);
  }

  return registry;
}

export function resolveReference(registry: DocumentRegistry, ref: string): DocumentRecord {
  const normalized = normalizeReference(ref);
  const matches: DocumentRecord[] = [];
  for (const index of [registry.byId, registry.byAlias, registry.byPath]) {
    const record = index.get(normalized);
    if (record && !matches.includes(record)) matches.push(record);
  }

  if (matches.length === 0 && looksLikePath(normalized)) {
    const pathRef = normalizePathAlias(normalized);
    const record = registry.byPath.get(pathRef) || registry.byPath.get(removePrefix(pathRef, 'content/'));
    if (record) matches.push(record);
  }

  if (matches.length === 0) throw new LookupError(`Document reference not found: ${ref}`);
  if (matches.length > 1) {
    const ids = matches.map((r) => r.id).sort().join(', ');
    throw new LookupError(`Document reference is ambiguous: ${ref} (${ids})`);
  }
  return matches[0];
}

/** Mirrors Python's `LookupError` for reference resolution. */
export class LookupError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'LookupError';
  }
}

export function normalizeReference(ref: string): string {
  let value = ref.trim();
  const wiki = WIKI_REF_RE.exec(value);
  if (wiki) value = wiki[1].trim();
  if (value.startsWith('doc:')) value = value.slice('doc:'.length).trim();
  return value.trim();
}

// ── Validation ────────────────────────────────────────────────────────────────

function validateRelatedDocs(records: DocumentRecord[], registry: DocumentRegistry): string[] {
  const violations: string[] = [];
  for (const record of records) {
    const sourceDir = posix.dirname(record.path);
    for (const related of record.related_docs) {
      const ref = normalizeReference(related);
      const candidates = relatedCandidates(ref, sourceDir);
      if (candidates.some((candidate) => canResolve(registry, candidate))) continue;
      violations.push(`${record.path}: related_docs reference not found: '${related}'`);
    }
  }
  return violations;
}

function validateAliasConflicts(
  byAlias: Map<string, DocumentRecord>,
  byId: Map<string, DocumentRecord>,
): string[] {
  const violations: string[] = [];
  for (const [alias, record] of byAlias) {
    const idRecord = byId.get(alias);
    if (idRecord && idRecord.path !== record.path) {
      violations.push(`alias '${alias}' from ${record.path} conflicts with id from ${idRecord.path}`);
    }
  }
  return violations;
}

function validateAliasPathConflicts(
  byAlias: Map<string, DocumentRecord>,
  byPath: Map<string, DocumentRecord>,
): string[] {
  const violations: string[] = [];
  for (const [alias, record] of byAlias) {
    const pathRecord = byPath.get(alias);
    if (pathRecord && pathRecord.path !== record.path) {
      violations.push(`alias '${alias}' from ${record.path} conflicts with path from ${pathRecord.path}`);
    }
  }
  return violations;
}

// ── Record construction ─────────────────────────────────────────────────────

function recordFromMetadata(
  repoPath: string,
  body: string,
  metadata: Frontmatter,
  mtime: number,
): DocumentRecord {
  const explicitId = asString(metadata.id);
  const docType = asString(metadata.doc_type) || inferDocType(repoPath);
  const docId = explicitId || generatedId(repoPath, docType);
  const stem = basename(repoPath).replace(/\.md$/, '');
  return {
    id: docId,
    aliases: asStrings(metadata.aliases),
    path: repoPath,
    title: asString(metadata.title) || firstHeading(body) || titleCase(stem.replace(/-/g, ' ')),
    doc_type: docType,
    summary: asString(metadata.summary),
    tags: asStrings(metadata.tags),
    systems: asStrings(metadata.systems),
    related_docs: asStrings(metadata.related_docs),
    updated_at: mtime,
    domain: inferDomain(repoPath),
    id_source: explicitId ? 'frontmatter' : 'generated',
  };
}

export function generatedId(repoPath: string, docType: string): string {
  const path = removeSuffix(removePrefix(repoPath, 'content/'), '.md');
  const parts = path.split('/').filter((part) => !ID_STRIP_DIRS.has(part));
  const slug = parts.join('.');
  return `${docType}.${slug}`.replace(/_/g, '-');
}

function addUnique(
  index: Map<string, DocumentRecord>,
  key: string,
  record: DocumentRecord,
  label: string,
  violations: string[],
): void {
  const existing = index.get(key);
  if (existing && existing.path !== record.path) {
    violations.push(`duplicate ${label} '${key}': ${existing.path} and ${record.path}`);
    return;
  }
  index.set(key, record);
}

// ── Reference helpers ─────────────────────────────────────────────────────────

function normalizeAlias(alias: string): string {
  if (looksLikePath(alias)) return normalizePathAlias(alias);
  return alias.trim();
}

function normalizePathAlias(alias: string): string {
  let path = alias.trim().replace(/\\/g, '/').replace(/^\/+/, '');
  if (!path.endsWith('.md')) throw new Error(alias);
  if (!path.startsWith('content/')) path = `content/${path}`;
  return path;
}

function relatedCandidates(ref: string, sourceDir: string): string[] {
  const candidates = [ref];
  if (looksLikePath(ref)) {
    const path = ref.trim().replace(/\\/g, '/').replace(/^\/+/, '');
    if (!path.startsWith('content/')) {
      candidates.push(`content/${path}`);
      candidates.push(posix.normalize(`${sourceDir}/${path}`));
    }
  }
  return candidates;
}

function canResolve(registry: DocumentRegistry, ref: string): boolean {
  try {
    resolveReference(registry, ref);
    return true;
  } catch {
    return false;
  }
}

function looksLikePath(value: string): boolean {
  return value.includes('/') || value.endsWith('.md');
}

// ── Frontmatter value coercion ────────────────────────────────────────────────

function asString(value: unknown): string {
  if (value === null || value === undefined) return '';
  if (Array.isArray(value)) return '';
  return String(value).trim();
}

function asStrings(value: unknown): string[] {
  if (value === null || value === undefined) return [];
  if (Array.isArray(value)) {
    return value
      .map((item) => stripQuotes(String(item).trim()))
      .filter((item) => item.length > 0);
  }
  if (typeof value === 'string') {
    const trimmed = value.trim();
    return trimmed ? [stripQuotes(trimmed)] : [];
  }
  return [String(value).trim()];
}

function stripQuotes(value: string): string {
  return value.replace(/^["']|["']$/g, '');
}

function firstHeading(body: string): string {
  for (const line of body.split('\n')) {
    if (line.startsWith('# ')) return line.slice(2).trim();
  }
  return '';
}

function inferDomain(path: string): string {
  const parts = path.split('/');
  if (parts.length >= 2 && parts[0] === 'content') return parts[1];
  return 'unknown';
}

function inferDocType(path: string): string {
  const parts = new Set(path.split('/'));
  if (parts.has('sops')) return 'sop';
  if (parts.has('templates')) return 'template';
  if (parts.has('reference')) return 'reference';
  if (parts.has('playbooks')) return 'playbook';
  if (parts.has('prompts')) return 'prompt';
  if (parts.has('archive')) return 'archive';
  return 'doc';
}

// ── Small string utilities (Python str method parity) ─────────────────────────

function removePrefix(value: string, prefix: string): string {
  return value.startsWith(prefix) ? value.slice(prefix.length) : value;
}

function removeSuffix(value: string, suffix: string): string {
  return value.endsWith(suffix) ? value.slice(0, -suffix.length) : value;
}

function titleCase(text: string): string {
  return text.replace(/[A-Za-z]+/g, (word) => word[0].toUpperCase() + word.slice(1).toLowerCase());
}

function basename(repoPath: string): string {
  return repoPath.split('/').pop() || repoPath;
}

function posixParent(absDir: string): string {
  const idx = absDir.lastIndexOf(sep);
  return idx > 0 ? absDir.slice(0, idx) : absDir;
}

function toPosixRel(file: string, repoRoot: string): string {
  return file.slice(repoRoot.length + 1).split(sep).join(posix.sep);
}

File: tools/content_tools/content_tools/doc_registry.py
from __future__ import annotations

import re
import posixpath
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from content_tools.sop_parse import parse_frontmatter, split_frontmatter


DOCUMENT_ID_RE = re.compile(r"^[a-z0-9][a-z0-9._-]*$")
WIKI_REF_RE = re.compile(r"^\[\[(?P<target>[^\]|]+)(?:\|[^\]]+)?\]\]$")
VALID_DOC_TYPES = {
    "sop",
    "checklist",
    "template",
    "reference",
    "playbook",
    "prompt",
    "task-template",
    "archive",
    "doc",
}


class DocumentRegistryError(ValueError):
    def __init__(self, violations: list[str]):
        self.violations = violations
        super().__init__("\n".join(violations))


@dataclass(frozen=True)
class DocumentRecord:
    id: str
    aliases: list[str]
    path: str
    title: str
    doc_type: str
    summary: str
    tags: list[str]
    systems: list[str]
    related_docs: list[str]
    updated_at: int
    domain: str
    id_source: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "stable_id": self.id_source == "frontmatter",
            "aliases": self.aliases,
            "path": self.path,
            "title": self.title,
            "doc_type": self.doc_type,
            "summary": self.summary,
            "tags": self.tags,
            "systems": self.systems,
            "related_docs": self.related_docs,
            "updated_at": self.updated_at,
            "domain": self.domain,
            "id_source": self.id_source,
        }


@dataclass(frozen=True)
class DocumentRegistry:
    documents: list[DocumentRecord]
    by_id: dict[str, DocumentRecord]
    by_alias: dict[str, DocumentRecord]
    by_path: dict[str, DocumentRecord]

    def to_dict(self) -> dict[str, Any]:
        return {"documents": [doc.to_dict() for doc in self.documents]}


def build_registry(content_root: Path, validate: bool = True) -> DocumentRegistry:
    root = content_root.resolve()
    repo_root = root.parent
    records: list[DocumentRecord] = []
    violations: list[str] = []

    for file_path in sorted(root.rglob("*.md")):
        text = file_path.read_text(encoding="utf-8", errors="replace")
        raw_frontmatter, body = split_frontmatter(text)
        metadata = parse_frontmatter(raw_frontmatter) if raw_frontmatter else {}
        repo_path = file_path.relative_to(repo_root).as_posix()
        record = _record_from_metadata(file_path, repo_path, body, metadata)
        records.append(record)

        if not DOCUMENT_ID_RE.match(record.id):
            violations.append(f"{repo_path}: invalid id {record.id!r}")
        for alias in record.aliases:
            if _looks_like_path(alias):
                try:
                    _normalize_path_alias(alias)
                except ValueError:
                    violations.append(f"{repo_path}: invalid path alias {alias!r}")
            elif not DOCUMENT_ID_RE.match(alias):
                violations.append(f"{repo_path}: invalid alias {alias!r}")
        if record.doc_type not in VALID_DOC_TYPES:
            violations.append(f"{repo_path}: unsupported doc_type {record.doc_type!r}")

    by_id: dict[str, DocumentRecord] = {}
    by_alias: dict[str, DocumentRecord] = {}
    by_path: dict[str, DocumentRecord] = {}

    for record in records:
        _add_unique(by_id, record.id, record, "id", violations)
        by_path[record.path] = record
        by_path[record.path.removeprefix("content/")] = record
        by_path["/" + record.path.removeprefix("content/")] = record
        for alias in record.aliases:
            normalized = _normalize_alias(alias)
            _add_unique(by_alias, normalized, record, "alias", violations)

    registry = DocumentRegistry(
        documents=records,
        by_id=by_id,
        by_alias=by_alias,
        by_path=by_path,
    )

    if validate:
        violations.extend(validate_alias_conflicts(by_alias, by_id))
        violations.extend(validate_alias_path_conflicts(by_alias, by_path))
        violations.extend(validate_related_docs(records, registry))
        if violations:
            raise DocumentRegistryError(violations)

    return registry


def resolve_reference(registry: DocumentRegistry, ref: str) -> DocumentRecord:
    normalized = normalize_reference(ref)
    matches: list[DocumentRecord] = []
    for index in (registry.by_id, registry.by_alias, registry.by_path):
        record = index.get(normalized)
        if record and record not in matches:
            matches.append(record)

    if not matches and _looks_like_path(normalized):
        path_ref = _normalize_path_alias(normalized)
        record = registry.by_path.get(path_ref) or registry.by_path.get(path_ref.removeprefix("content/"))
        if record:
            matches.append(record)

    if not matches:
        raise LookupError(f"Document reference not found: {ref}")
    if len(matches) > 1:
        ids = ", ".join(sorted(record.id for record in matches))
        raise LookupError(f"Document reference is ambiguous: {ref} ({ids})")
    return matches[0]


def normalize_reference(ref: str) -> str:
    value = ref.strip()
    wiki = WIKI_REF_RE.match(value)
    if wiki:
        value = wiki.group("target").strip()
    if value.startswith("doc:"):
        value = value.removeprefix("doc:").strip()
    return value.strip()


def validate_related_docs(records: list[DocumentRecord], registry: DocumentRegistry) -> list[str]:
    violations: list[str] = []
    for record in records:
        source_dir = str(Path(record.path).parent)
        for related in record.related_docs:
            ref = normalize_reference(related)
            candidates = _related_candidates(ref, source_dir)
            if any(_can_resolve(registry, candidate) for candidate in candidates):
                continue
            violations.append(f"{record.path}: related_docs reference not found: {related!r}")
    return violations


def validate_alias_conflicts(
    by_alias: dict[str, DocumentRecord],
    by_id: dict[str, DocumentRecord],
) -> list[str]:
    violations: list[str] = []
    for alias, record in by_alias.items():
        id_record = by_id.get(alias)
        if id_record and id_record.path != record.path:
            violations.append(f"alias {alias!r} from {record.path} conflicts with id from {id_record.path}")
    return violations


def validate_alias_path_conflicts(
    by_alias: dict[str, DocumentRecord],
    by_path: dict[str, DocumentRecord],
) -> list[str]:
    violations: list[str] = []
    for alias, record in by_alias.items():
        path_record = by_path.get(alias)
        if path_record and path_record.path != record.path:
            violations.append(f"alias {alias!r} from {record.path} conflicts with path from {path_record.path}")
    return violations


def _record_from_metadata(
    file_path: Path,
    repo_path: str,
    body: str,
    metadata: dict[str, Any],
) -> DocumentRecord:
    explicit_id = _string(metadata.get("id"))
    doc_type = _string(metadata.get("doc_type")) or _infer_doc_type(repo_path)
    doc_id = explicit_id or _generated_id(repo_path, doc_type)
    return DocumentRecord(
        id=doc_id,
        aliases=_strings(metadata.get("aliases")),
        path=repo_path,
        title=_string(metadata.get("title")) or _first_heading(body) or file_path.stem.replace("-", " ").title(),
        doc_type=doc_type,
        summary=_string(metadata.get("summary")),
        tags=_strings(metadata.get("tags")),
        systems=_strings(metadata.get("systems")),
        related_docs=_strings(metadata.get("related_docs")),
        updated_at=int(file_path.stat().st_mtime),
        domain=_infer_domain(repo_path),
        id_source="frontmatter" if explicit_id else "generated",
    )


def _generated_id(repo_path: str, doc_type: str) -> str:
    path = repo_path.removeprefix("content/").removesuffix(".md")
    parts = [part for part in path.split("/") if part not in {"sops", "templates", "reference", "playbooks"}]
    slug = ".".join(parts)
    return f"{doc_type}.{slug}".replace("_", "-")


def _add_unique(
    index: dict[str, DocumentRecord],
    key: str,
    record: DocumentRecord,
    label: str,
    violations: list[str],
) -> None:
    existing = index.get(key)
    if existing and existing.path != record.path:
        violations.append(f"duplicate {label} {key!r}: {existing.path} and {record.path}")
        return
    index[key] = record


def _normalize_alias(alias: str) -> str:
    if _looks_like_path(alias):
        return _normalize_path_alias(alias)
    return alias.strip()


def _normalize_path_alias(alias: str) -> str:
    path = alias.strip().replace("\\", "/").lstrip("/")
    if not path.endswith(".md"):
        raise ValueError(alias)
    if not path.startswith("content/"):
        path = f"content/{path}"
    return path


def _related_candidates(ref: str, source_dir: str) -> list[str]:
    candidates = [ref]
    if _looks_like_path(ref):
        path = ref.strip().replace("\\", "/").lstrip("/")
        if not path.startswith("content/"):
            candidates.append(f"content/{path}")
            candidates.append(posixpath.normpath(f"{source_dir}/{path}"))
    return candidates


def _can_resolve(registry: DocumentRegistry, ref: str) -> bool:
    try:
        resolve_reference(registry, ref)
        return True
    except LookupError:
        return False
    except ValueError:
        return False


def _looks_like_path(value: str) -> bool:
    return "/" in value or value.endswith(".md")


def _string(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, list):
        return ""
    return str(value).strip()


def _strings(value: Any) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(item).strip().strip('"').strip("'") for item in value if str(item).strip()]
    if isinstance(value, str):
        return [value.strip().strip('"').strip("'")] if value.strip() else []
    return [str(value).strip()]


def _first_heading(body: str) -> str:
    for line in body.splitlines():
        if line.startswith("# "):
            return line.removeprefix("# ").strip()
    return ""


def _infer_domain(path: str) -> str:
    parts = path.split("/")
    if len(parts) >= 2 and parts[0] == "content":
        return parts[1]
    return "unknown"


def _infer_doc_type(path: str) -> str:
    parts = set(path.split("/"))
    if "sops" in parts:
        return "sop"
    if "templates" in parts:
        return "template"
    if "reference" in parts:
        return "reference"
    if "playbooks" in parts:
        return "playbook"
    if "prompts" in parts:
        return "prompt"
    if "archive" in parts:
        return "archive"
    return "doc"
